File: lib/scala/filewatcher/src/main/scala/org/enso/filewatcher/WatcherFactory.scala
package org.enso.filewatcher

import java.nio.file.Path

trait WatcherFactory {

  /** Create a instance of [[Watcher]].
    *
    * @param root the directory to watch
    * @param eventCallback the callback handling the file watcher events
    * @param exceptionCallback the callback handling the file watcher errors
    * @return a new instance of [[Watcher]]
    */
  def build(
    root: Path,
    eventCallback: Watcher.WatcherEvent => Unit,
    exceptionCallback: Watcher.WatcherError => Unit
  ): Watcher
}

File: lib/scala/filewatcher/src/main/java/org/enso/filewatcher/WatcherFactory.java
package org.enso.filewatcher;

import java.nio.file.Path;
import java.util.function.Consumer;

/** Factory for directory watcher */
public interface WatcherFactory {

  static WatcherFactory createDefault() {
    return new DefaultWatcherFactory();
  }

  /**
   * @param root Must be existing directory
   */
  Watcher build(
      Path root,
      Consumer<Watcher.WatcherEvent> eventCallback,
      Consumer<Watcher.WatcherError> exceptionCallback);
}
